api/types/clusterconfig.go

```diff
@@ -322,16 +334,73 @@ func (c *ClusterConfigV3) CheckAndSetDefaults() error {
 		c.Spec.KeepAliveCountMax = int64(defaults.KeepAliveCountMax)
 	}
 
+	// Detect invalid or inconsistent overrides.
+	if err := c.keepOnlyValidOverrides(); err != nil {
+		return trace.Wrap(err)
+	}
+
 	return nil
 }
 
+// keepOnlyValidOverrides checks for inconsistencies with the current override
+// profile, and resets inadmissible overrides so that they do not impact the
+// value of the corresponding ClusterConfig field.
+func (c *ClusterConfigV3) keepOnlyValidOverrides() error {
+	var errs []error
+	if !c.HasOverride() {
+		return nil
+	}
+
+	if c.Override.SessionRecording != "" {
+		var err error
+		allowedOverrides := []string{RecordAtNode, RecordAtNodeSync, RecordOff}
+		if !utils.SliceContainsStr(allowedOverrides, c.Override.SessionRecording) {
+			err = trace.BadParameter("session_recording can be overriden only with one of: %v", strings.Join(allowedOverrides, ","))
+		}
+		if !utils.SliceContainsStr(allowedOverrides, c.Spec.SessionRecording) {
+			err = trace.BadParameter("session_recording=%q cannot be overriden", c.Spec.SessionRecording)
+		}
+		if err != nil {
+			errs = append(errs, err)
+			c.Override.SessionRecording = ""
```

Comment:
Why do you need to empty the field here? Wouldn't the error just be returned to the user when they tried to apply invalid override? If this is indeed needed, add a comment explaining why.

Reply:
It's needed to support partial overrides, i.e. keeping as much of an override as is applicable, while dropping/zeroing those parts that aren't. I'll add a comment.